ec/ecresource/deploymentresource/apm/v2/apm.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/elastic/terraform-provider-ec/ec/internal/util"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+type ApmTF struct {
+	ElasticsearchClusterRefId types.String `tfsdk:"elasticsearch_cluster_ref_id"`
+	RefId                     types.String `tfsdk:"ref_id"`
+	ResourceId                types.String `tfsdk:"resource_id"`
+	Region                    types.String `tfsdk:"region"`
+	HttpEndpoint              types.String `tfsdk:"http_endpoint"`
+	HttpsEndpoint             types.String `tfsdk:"https_endpoint"`
+	InstanceConfigurationId   types.String `tfsdk:"instance_configuration_id"`
+	Size                      types.String `tfsdk:"size"`
+	SizeResource              types.String `tfsdk:"size_resource"`
+	ZoneCount                 types.Int64  `tfsdk:"zone_count"`
+	Config                    types.Object `tfsdk:"config"`
+}
+
+type Apm struct {
+	ElasticsearchClusterRefId *string    `tfsdk:"elasticsearch_cluster_ref_id"`
+	RefId                     *string    `tfsdk:"ref_id"`
+	ResourceId                *string    `tfsdk:"resource_id"`
+	Region                    *string    `tfsdk:"region"`
+	HttpEndpoint              *string    `tfsdk:"http_endpoint"`
+	HttpsEndpoint             *string    `tfsdk:"https_endpoint"`
+	InstanceConfigurationId   *string    `tfsdk:"instance_configuration_id"`
+	Size                      *string    `tfsdk:"size"`
+	SizeResource              *string    `tfsdk:"size_resource"`
+	ZoneCount                 int        `tfsdk:"zone_count"`
+	Config                    *ApmConfig `tfsdk:"config"`
+}
+
+func ReadApms(in []*models.ApmResourceInfo) (*Apm, error) {
+	for _, model := range in {
+		if util.IsCurrentApmPlanEmpty(model) || utils.IsApmResourceStopped(model) {
+			continue
+		}
+
+		apm, err := ReadApm(model)
+		if err != nil {
+			return nil, err
+		}
+
+		return apm, nil
+	}
+
+	return nil, nil
+}
+
+func ReadApm(in *models.ApmResourceInfo) (*Apm, error) {
+	var apm Apm
+
+	apm.RefId = in.RefID
+
+	apm.ResourceId = in.Info.ID
+
+	apm.Region = in.Region
+
+	plan := in.Info.PlanInfo.Current.Plan
+
+	topologies, err := ReadApmTopologies(plan.ClusterTopology)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(topologies) > 0 {
+		apm.InstanceConfigurationId = topologies[0].InstanceConfigurationId
+		apm.Size = topologies[0].Size
+		apm.SizeResource = topologies[0].SizeResource
+		apm.ZoneCount = topologies[0].ZoneCount
+	}
+
+	apm.ElasticsearchClusterRefId = in.ElasticsearchClusterRefID
+
+	apm.HttpEndpoint, apm.HttpsEndpoint = converters.ExtractEndpoints(in.Info.Metadata)
+
+	configs, err := readApmConfigs(plan.Apm)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(configs) > 0 {
+		apm.Config = &configs[0]
+	}
+
+	return &apm, nil
+}
+
+func (apm ApmTF) Payload(ctx context.Context, payload models.ApmPayload) (*models.ApmPayload, diag.Diagnostics) {
```

Comment:
Should be called something like `EnrichTemplate`

Reply:
My take was that the func generates payload for Apm based on initial payload from a template. It doesn't matter how the func does it.